File: src/main/java/com/somefrills/features/core/ToggleFeature.java
package com.somefrills.features.core;

import com.somefrills.misc.KeybindManager;
import io.github.notenoughupdates.moulconfig.observer.Property;

public abstract class ToggleFeature extends AbstractFeature {
    private final Property<Integer> keybindProperty;

    private KeybindManager.Subscription sub = null;
    private boolean keybindActive = false;

    public ToggleFeature(Property<Boolean> enabledProperty,
                         Property<Integer> keybindProperty) {
        super(enabledProperty);
        this.keybindProperty = keybindProperty;
    }

    @Override
    protected boolean evaluate() {
        return keybindActive;
    }

    @Override
    public final void toggle() {
        if (!isEnabled()) return;

        keybindActive = !keybindActive;
        sync();
    }

    @Override
    protected final void onEnable() {
        if (sub != null) {
            sub.unregister();
            sub = null;
        }
        sub = KeybindManager.register(keybindProperty, this::toggle);
    }

    @Override
    protected final void onDisable() {
        if (sub != null) {
            sub.unregister();
            sub = null;
        }
        keybindActive = false;
    }
}

File: src/main/kotlin/com/somefrills/features/core/ToggleFeature.kt
package com.somefrills.features.core

import com.somefrills.misc.KeybindManager
import com.somefrills.misc.KeybindManager.register
import io.github.notenoughupdates.moulconfig.observer.Property

abstract class ToggleFeature(
    enabledProperty: Property<Boolean>,
    private val keybindProperty: Property<Int>
) : AbstractFeature(enabledProperty) {
    private var sub: KeybindManager.Subscription? = null
    private var keybindActive = false

    override fun evaluate(): Boolean {
        return keybindActive
    }

    override fun toggle() {
        if (!isEnabled()) return

        keybindActive = !keybindActive
        sync()
    }

    override fun onEnable() {
        if (sub != null) {
            sub!!.unregister()
            sub = null
        }
        sub = register(keybindProperty, Runnable { this.toggle() })
    }

    override fun onDisable() {
        if (sub != null) {
            sub!!.unregister()
            sub = null
        }
        keybindActive = false
    }
}